src/security/builtin_plugins/authentication/src/authentication.c

```diff
@@ -650,6 +655,45 @@ static DDS_Security_ValidationResult_t get_adjusted_participant_guid(X509 *cert,
 }
 #undef ADJUSTED_GUID_PREFIX_FLAG
 
+
+#if OPENSSL_VERSION_NUMBER < 0x30000000L
+static bool load_pkcs11_provider(void *arg)
+{
+  dds_security_authentication_impl *auth = arg;
+  const char *pkcs11 = "pkcs11";
+  bool result = true;
+
+  ddsrt_mutex_lock(&auth->lock);
+  if (!auth->engine)
+  {
+    if (!(auth->engine = ENGINE_by_id(pkcs11)))
+      result = false;
+    else if (!ENGINE_init(auth->engine))
+    {
+      ENGINE_free(auth->engine);
+      auth->engine = NULL;
+      result = false;
+    }
+  }
+  ddsrt_mutex_unlock(&auth->lock);
+  return result;
+}
+#else
+static bool load_pkcs11_provider(void *arg)
+{
+  dds_security_authentication_impl *auth = arg;
+  const char *pkcs11 = "pkcs11";
+
+  if (OSSL_PROVIDER_available(NULL, pkcs11))
+    return true;
+
+  ddsrt_mutex_lock(&auth->lock);
+  auth->pkcs11Provider = OSSL_PROVIDER_try_load(NULL, pkcs11, 1);
+  ddsrt_mutex_unlock(&auth->lock);
+  return (auth->pkcs11Provider != NULL);
```

Comment:
If `pkcs11Provider` is checked outside the lock, then why do we need the mutex here in the first place?